src/string.cr

```diff
@@ -2864,7 +2864,8 @@ class String
         yield str, match, buffer
 
         if str.bytesize == 0
-          byte_offset = index + 1
+          # The pattern matched an empty result. We must advance one character to avoid stagnation.
+          byte_offset = index + Char::Reader.new(self, pos: byte_offset).current_char_width
```

Comment:
Actually, `#gsub(string : String, &block)` has the same issue, although it is probably not a regression:

```crystal
"×".gsub("", "÷") # => "÷\xC3÷\x97÷"
```

Reply:
I was planning to go through all string operations to look for suspicious `+ 1` 😆
But this one surely makes sense to fix in the same go.